go/mysql/binlog_event_compression.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/klauspost/compress/zstd"
+
+	"vitess.io/vitess/go/vt/log"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// This file contains code related to handling compression related
+// events. More specifically today, compressed transaction payloads:
+// See: https://dev.mysql.com/doc/refman/en/binary-log-transaction-compression.html
+
+// Transaction Payload wire protocol fields:
+// https://dev.mysql.com/doc/dev/mysql-server/latest/classbinary__log_1_1codecs_1_1binary_1_1Transaction__payload.html
+const (
+	payloadHeaderEndMark = iota
+	payloadSizeField
+	payloadCompressionTypeField
+	payloadUncompressedSizeField
+)
+
+// Compression algorithms that are supported (only zstd today
+// in MySQL 8.0):
+// https://dev.mysql.com/doc/refman/8.0/en/binary-log-transaction-compression.html
+const (
+	transactionPayloadCompressionZstd = 0
+	transactionPayloadCompressionNone = 255
+)
+
+var transactionPayloadCompressionTypes = map[uint64]string{
+	transactionPayloadCompressionZstd: "ZSTD",
+	transactionPayloadCompressionNone: "NONE",
+}
+
+// Create a reader that caches decompressors.
+var zstdDecoder, _ = zstd.NewReader(nil, zstd.WithDecoderConcurrency(0))
+
+type TransactionPayload struct {
+	Size             uint64
+	CompressionType  uint64
+	UncompressedSize uint64
```

Comment:
Can this be so large it needs a `uint64` or is MySQL limited to lower values like 4GB that are `uint32` encoded? Asking since other places use `uint32` like the log position in a binlog event. Based on https://dev.mysql.com/doc/refman/8.0/en/binary-log-transaction-compression.html, the implied compressed limit at least seems to be 1GB?

Reply:
The compressed size would be limited by max_allowed_packet and thus 1GB. I used uint64 everywhere as that's what libbinlogevents was doing in MySQL. That and virtually all of the platforms vitess runs on will have a 64 bit word size. I can change it though if you think it's problematic for any reason(s).

Reply:
@mattlord I think `uint64` is fine here, was mostly wondering here. Since we recently cleaned up some inconsistent typing but if MySQL itself does also uint64 that's most consistent I think.